### PR TITLE
Server Tutorial 3 - JSON

### PR DESCRIPTION
**Pull request by sehhajdkg**

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -19,14 +19,38 @@
...
 public class DataServlet extends HttpServlet {
 
   @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello Sehhaj!</h1>");
+  public void doGet(final HttpServletRequest request, final HttpServletResponse response) throws IOException {
+
+        // Initialise messages with three hard-coded values
+        ArrayList<String> messages = new ArrayList<String>();
```
Comment (lorikeet-s) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Use List<String> on the declaration side of the expression. Unless you're actually relying on some specific features that ArrayList gives you it's better to leave that to the implementation details (the new ArrayList<String> side of things).
(This tends to be more important in a class interface than a local variable but it's a good practice to get into)

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -19,14 +19,38 @@
...
+        // Send JSON as response
+        response.setContentType("application/json;");
+        response.getWriter().println(json);
+    }
+
+    /**
+     * Converts a an ArrayList<String> instance into a JSON string using the Gson
+     * library.
+     */
+    private String convertToJsonUsingGson(ArrayList<String> data) {
```
Comment (lorikeet-s) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: same here with List vs ArrayList, it's more important here.

**Review by lorikeet-s — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -19,14 +19,38 @@
...
+        final Gson gson = new Gson();
```
Comment (g-wittie) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: You might want to make Gson a private variable of the class. Right now you're creating a new Gson object every time the function is called.

It's probably relatively expensive to instantiate and it looks like it doesn't have state, so if you move it to a private var it can be reused for all calls to this function.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 1 above)
Comment (sehhajdkg) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Done

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 2 above)
Comment (sehhajdkg) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Done :)

Comment (g-wittie) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Two things here:
1) You don't need to state that it converts 'an ArrayList<String>' since the method signature tells us that. Comments shouldn't repeat information or state the obvious. Both the comment and the function name also omit something that could be of interest to the caller: the returned value isn't just a JSON string, it's a JSON array.

2) Both the comment and the method signature mention that the method uses the Gson library. This is what we call an 'implementation detail'. The caller shouldn't need to worry about what kind of library you use to do the conversion. The important thing is that the output is a json array.

With those things in mind I'd reword these to be something along the lines of:
```
// Converts the input to a JSON array.
private String toJsonArray(List<String> items)
```

Another thing I've done is omit the word 'convert'. This strays more into preference territory, but you want to keep function / method names as short as possible while still being descriptive, and you want to avoid repeating information. With that in mind, in this context I think a reader will understand 'convertToX' and 'toX' to mean the same thing, so the shorter form is preferable.

**Review by g-wittie — CHANGES_REQUESTED** [review]

**Review by g-wittie — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 3 above)
Comment (sehhajdkg) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Done :)

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 2 above)
Comment (sehhajdkg) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Done :)

**Review by lorikeet-s — APPROVED** [review]
